Source/WebCore/css/calc/CSSCalcTree+Parser.cpp

```diff
@@ -703,6 +710,192 @@ static std::optional<TypedChild> consumeProgress(CSSParserTokenRange& tokens, in
+        return std::nullopt;
+    }
+
+    if (!CSSPropertyParserHelpers::consumeCommaIncludingWhitespace(tokens)) {
+        LOG_WITH_STREAM(Calc, stream << "Failed '" << nameLiteralForSerialization(Op::id) << "' function - missing comma");
+        return std::nullopt;
+    }
+
+    auto end = parseCalcSum(tokens, depth, nestedState);
+    if (!end) {
+        LOG_WITH_STREAM(Calc, stream << "Failed '" << nameLiteralForSerialization(Op::id) << "' function - failed parse of argument #3");
+        return std::nullopt;
+    }
+
+    if (!tokens.atEnd()) {
+        LOG_WITH_STREAM(Calc, stream << "Failed '" << nameLiteralForSerialization(Op::id) << "' function - extraneous tokens found");
+        return std::nullopt;
+    }
+
+    // - Validate arguments
+
+    if (!Type::consistentType(start->type, end->type)) {
+        LOG_WITH_STREAM(Calc, stream << "Failed '" << nameLiteralForSerialization(Op::id) << "' function - inconsistent types");
+        return std::nullopt;
+    }
+
+    if (!start->type.matches(schemaCategory)) {
+        LOG_WITH_STREAM(Calc, stream << "Failed '" << nameLiteralForSerialization(Op::id) << "' function - inconsistent types");
+        return std::nullopt;
+    }
+
+    // `media-progress() always evaluates to a <number>.
+    auto outputType = Type { };
+
+    Op op { schema, WTFMove(start->child), WTFMove(end->child) };
+
+    if (auto* simplificationOptions = state.simplificationOptions) {
+        if (auto replacement = simplify(op, *simplificationOptions))
+            return TypedChild { WTFMove(*replacement), outputType };
+    }
+    return TypedChild { makeChild(WTFMove(op), outputType), outputType };
+}
+
+static std::optional<TypedChild> consumeContainerProgress(CSSParserTokenRange& tokens, int depth, ParserState& state)
+{
+    // <container-progress()> = container-progress( <mf-name> [ of <container-name> ]?, <calc-sum>, <calc-sum> )
+
+    if (!state.parserContext.cssContainerProgressFunctionEnabled)
+        return { };
+
+    using Op = ContainerProgress;
+
+    auto featureName = consumeMediaFeatureName(tokens);
+    if (!featureName) {
+        LOG_WITH_STREAM(Calc, stream << "Failed '" << nameLiteralForSerialization(Op::id) << "' function - failed parse of argument #1");
+        return std::nullopt;
+    }
+
+    AtomString container;
+    if (CSSPropertyParserHelpers::consumeIdentRaw<CSSValueOf>(tokens)) {
+        if (tokens.peek().type() != IdentToken || !isValidCustomIdentifier(tokens.peek().id()) || !CSSPropertyParserHelpers::isValidContainerNameIdentifier(tokens.peek().id())) {
+            LOG_WITH_STREAM(Calc, stream << "Failed '" << nameLiteralForSerialization(Op::id) << "' function - failed parse of argument #1");
+            return std::nullopt;
+        }
+        container = tokens.consumeIncludingWhitespace().value().toAtomString();
+    }
+
+    const auto* schema = CQ::ContainerQueryParser::containerProgressProvidingSchemaForFeatureName(*featureName, state.parserContext);
```

Comment:
Is `const auto*` really better than `auto*` or `auto` here? Presumably the const-ness is already present in the return value and isn’t added by stating the type?